Add list of enabled modules to welcome message

dimitrit opened the pull request:
Fixes #31

dimitrit committed:
Add list of enabled modules to welcome message

Fixes #31

marcelstoer commented:
Looks just fine, thanks. Can you provide a sample boot message from the firmware you tested with?

dimitrit committed:
Allow for leading white-space in user_modules.h

Modules with leading white-space are included in the list of names when enabled (cf. TCS34725).

marcelstoer commented on cmd.sh:
```diff
@@ -57,7 +57,7 @@ cd /opt/nodemcu-firmware
 # I never found a good solution
 if [ "$CAN_MODIFY_VERSION" = true ]; then
-  MODULES=$(awk '/^#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -)
+  MODULES=$(awk '/^[ \t]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -)
```
How about using the POSIX `[:space:]` for that like in the line below.

dimitrit replied:
Awk doesn't seem to like `[:space:]`...I'll have a look to see why!

dimitrit replied:
gawk(1) seems to work, but that means adding it to the list of packages to apt-get in the Dockerfile.  Let me know which you prefer (ie. leaving it as `[ \t]`  using awk, or changing it to `[:space:]` and adding gawk).

marcelstoer replied:
The shorthand equivalent for `[:space:]` is `\s` and my macOS `awk` is happy with that.

```
~/Data/NodeMCU/nodemcu-firmware (dev) > awk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tls,tmr,uart,wifi
```
It's a detail but `\s*` is IMO preferable to `[ \t]*` because it matches all whitespace and not just spaces and tabs (however unlikely it is to find those).

dimitrit replied:
With `LUA_USE_MODULES_TCS34725` enabled on my Mac:
```
preston:nodemcu-firmware dimitrit$ awk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tls,tmr,uart,wifi
preston:nodemcu-firmware dimitrit$ awk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
```
Which implies that `\s` and `[:space:]` are not actually equivalent in macOS.

However, in Ubuntu they do appear to behave the same:
```
preston:nodemcu-firmware dimitrit$ docker run --rm -ti -v `pwd`:/opt/nodemcu-firmware docker-nodemcu-build /bin/bash
root@b9860b50cfe4:/opt/nodemcu-firmware# awk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
root@b9860b50cfe4:/opt/nodemcu-firmware# awk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
root@b9860b50cfe4:/opt/nodemcu-firmware# exit
```

@marcelstoer, please could you check the above on your Mac? In any case, given that the docker image is based on Ubuntu I agree that it makes sense to change the regex to `[:space:]`.

[UPDATE]: It seems I accidentally used the 'wrong' image when I did the above tests in Ubuntu (i.e. one that had `gawk` installed) - see my comment below!

dimitrit replied:
Just did a few more tests...it turns out that in the Ubuntu Docker image `awk` is a symbolic link, which (finally) links to `mawk`:

```
root@d717ed48657b:/opt/nodemcu-firmware# ls -l /usr/bin/awk
lrwxrwxrwx 1 root root 21 Feb 28 19:13 /usr/bin/awk -> /etc/alternatives/awk
root@d717ed48657b:/opt/nodemcu-firmware# ls -l /etc/alternatives/awk
lrwxrwxrwx 1 root root 13 Feb 28 19:13 /etc/alternatives/awk -> /usr/bin/mawk
root@d717ed48657b:/opt/nodemcu-firmware# file /usr/bin/mawk
/usr/bin/mawk: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.24, BuildID[sha1]=4e43393facf439efde422d29048db342528974ad, stripped
```

However, when `gawk` is installed,  `awk` is linked to `gawk`:
```
root@7beef77f07c1:/opt/nodemcu-firmware# which awk
/usr/bin/awk
root@7beef77f07c1:/opt/nodemcu-firmware# ls -l /usr/bin/awk
lrwxrwxrwx 1 root root 21 Feb 28 19:13 /usr/bin/awk -> /etc/alternatives/awk
root@7beef77f07c1:/opt/nodemcu-firmware# ls -l /etc/alternatives/awk
lrwxrwxrwx 1 root root 13 Apr  7 16:58 /etc/alternatives/awk -> /usr/bin/gawk
root@7beef77f07c1:/opt/nodemcu-firmware# file /usr/bin/gawk
/usr/bin/gawk: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=38028e7fa7d204e30764568640ead5bb665aa125, stripped
```

It appears that `mawk` does not support `[:space:]`, and `\s` does not match leading white space, whereas `gawk` works just fine:
```
root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/mawk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tls,tmr,uart,wifi
root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/mawk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -

root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/gawk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/gawk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
```

It seems to me that the best solution will be to include `gawk` in list of packages to `apt-get` in the Dockerfile; `gawk` does not appear to impact the actual firmware build, and will enable use of POSIX expressions.

@marcelstoer I have already tested this on my Mac, but not yet pushed the changes pending your feedback.

```
NodeMCU 2.2.0.0 built with Docker provided by frightanic.com
	branch: master
	commit: 67027c0d05f7e8d1b97104e05a3715f6ebc8d07f
	SSL: false
	modules: adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
 build created on 2018-04-08 04:13
 powered by Lua 5.1.4 on SDK 2.2.1(cfd48f3)
lua: cannot open init.lua
> 
```

marcelstoer replied:
Woah, all those subtle differences are a real pain; quite a disappointment to be honest. Thanks for testing. I wouldn't want to include `gawk` just to find out that with the next version the behavior changed again. As `\s` (and `[:space:]` for that matter) matches `\r\n\v\f` in addition to space and tab and as those should _really_ never be part of our search space I opt to stick with your original proposal.

dimitrit replied:
Agree...those subtleties made this far more painful than it should have been.  

Thanks for merging! :D

dimitrit committed:
Show a dash when no user modules are enabled

dimitrit commented:
Here are a couple of samples showing the module names in the boot message:

* Random selection of user modules, including ones with a comment (ENDUSER_SETUP) and leading white-space (TCS34725):  
```
NodeMCU 2.2.0.0 built with Docker provided by frightanic.com
	branch: master
	commit: 67027c0d05f7e8d1b97104e05a3715f6ebc8d07f
	SSL: false
	modules: bit,enduser_setup,file,gpio,net,node,tcs34725,uart,wifi
 build created on 2018-04-07 18:07
 powered by Lua 5.1.4 on SDK 2.2.1(cfd48f3)
lua: cannot open init.lua
> 
```
  
* A dash is shown when no user modules are enabled:  
```
NodeMCU 2.2.0.0 built with Docker provided by frightanic.com
	branch: master
	commit: 67027c0d05f7e8d1b97104e05a3715f6ebc8d07f
	SSL: false
	modules: -
 build created on 2018-04-07 18:14
 powered by Lua 5.1.4 on SDK 2.2.1(cfd48f3)
lua: cannot open init.lua
> 
```

marcelstoer commented:
Thanks a lot Dimitri for your dedication and the thorough testing!
